In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv('LoanApprovalPrediction.csv')

In [3]:
df

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

In [4]:
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [5]:
df.tail()

Loan_ID  Gender Married Dependents Education Self_Employed  \
609  LP002978  Female      No          0  Graduate            No   
610  LP002979    Male     Yes         3+  Graduate            No   
611  LP002983    Male     Yes          1  Graduate            No   
612  LP002984    Male     Yes          2  Graduate            No   
613  LP002990  Female      No          0  Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N

In [6]:
df.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [7]:
df.info

<bound method DataFrame.info of       Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849          

In [8]:
df.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  
count      564.000000  
mean         0.842199  
std          0.364878  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [9]:
df.drop(['Loan_ID'],axis=1,inplace=True)
df

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0      Male      No          0      Graduate            No             5849   
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0         NaN             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
0           Urban           Y  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[614 rows x 12 columns]

In [10]:
df.isna().sum()

Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [11]:
df['Gender'].fillna(df['Gender'].mode()[0],inplace=True)
df['Married'].fillna(df['Gender'].mode()[0],inplace=True)
df['Dependents'].fillna(df['Gender'].mode()[0],inplace=True)
df['Self_Employed'].fillna(df['Self_Employed'].mode()[0],inplace=True)
df['LoanAmount'].fillna(df['LoanAmount'].mode()[0],inplace=True)
df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].mode()[0],inplace=True)
df['Credit_History'].fillna(df['Credit_History'].mode()[0],inplace=True)

In [12]:
df.isna().sum()

Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

In [13]:
df.dtypes

Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

In [14]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
for i in df.select_dtypes(include=['object']):
    df[i]=le.fit_transform(df[i])

In [15]:
df.dtypes

Gender                 int32
Married                int32
Dependents             int32
Education              int32
Self_Employed          int32
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int32
Loan_Status            int32
dtype: object

In [16]:
df

Gender  Married  Dependents  Education  Self_Employed  ApplicantIncome  \
0         1        1           0          0              0             5849   
1         1        2           1          0              0             4583   
2         1        2           0          0              1             3000   
3         1        2           0          1              0             2583   
4         1        1           0          0              0             6000   
..      ...      ...         ...        ...            ...              ...   
609       0        1           0          0              0             2900   
610       1        2           3          0              0             4106   
611       1        2           1          0              0             8072   
612       1        2           2          0              0             7583   
613       0        1           0          0              1             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0       120.0             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

     Property_Area  Loan_Status  
0                2            1  
1                0            0  
2                2            1  
3                2            1  
4                2            1  
..             ...          ...  
609              0            1  
610              0            1  
611              2            1  
612              2            1  
613              1            0  

[614 rows x 12 columns]

In [17]:
x=df.iloc[:,:-1].values
y=df.iloc[:,-1].values

In [18]:
x.ndim

2

In [19]:
y.ndim

1

In [20]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.30,random_state=42)
x_train

array([[  1.,   2.,   0., ..., 360.,   0.,   2.],
       [  1.,   2.,   0., ..., 360.,   1.,   0.],
       [  1.,   1.,   0., ..., 360.,   1.,   1.],
       ...,
       [  0.,   1.,   0., ..., 360.,   1.,   2.],
       [  0.,   0.,   4., ..., 240.,   1.,   1.],
       [  1.,   2.,   4., ..., 360.,   1.,   2.]])

In [21]:
x_train.shape

(429, 11)

In [22]:
x_test

array([[  1.,   2.,   0., ..., 360.,   1.,   1.],
       [  1.,   2.,   0., ..., 360.,   1.,   1.],
       [  1.,   2.,   2., ..., 360.,   1.,   0.],
       ...,
       [  1.,   2.,   2., ..., 180.,   1.,   2.],
       [  1.,   2.,   0., ..., 360.,   1.,   0.],
       [  0.,   2.,   0., ..., 360.,   0.,   0.]])

In [23]:
x_test.shape

(185, 11)

In [24]:
from sklearn.preprocessing import StandardScaler
scalar=StandardScaler()
scalar.fit(x_train)
x_train=scalar.fit_transform(x_train)
x_test=scalar.fit_transform(x_test)
x_train

array([[ 0.49343516,  0.77517165, -0.73209889, ...,  0.30437507,
        -2.47991935,  1.20651188],
       [ 0.49343516,  0.77517165, -0.73209889, ...,  0.30437507,
         0.40323892, -1.34950589],
       [ 0.49343516, -1.26500404, -0.73209889, ...,  0.30437507,
         0.40323892, -0.071497  ],
       ...,
       [-2.02660871, -1.26500404, -0.73209889, ...,  0.30437507,
         0.40323892,  1.20651188],
       [-2.02660871, -3.30517972,  2.84705123, ..., -1.45542149,
         0.40323892, -0.071497  ],
       [ 0.49343516,  0.77517165,  2.84705123, ...,  0.30437507,
         0.40323892,  1.20651188]])

In [28]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
model1=KNeighborsClassifier()
model2=GaussianNB()
model3=SVC()
model4=DecisionTreeClassifier()
model5=RandomForestClassifier()
models=[model1,model2,model3,model4,model5]

In [29]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, ConfusionMatrixDisplay

for model in models:
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    print(model)
    print(classification_report(y_test, y_pred))
    print(confusion_matrix(y_test, y_pred))
    print(accuracy_score(y_test, y_pred))

KNeighborsClassifier()
              precision    recall  f1-score   support

           0       0.76      0.43      0.55        65
           1       0.75      0.93      0.83       120

    accuracy                           0.75       185
   macro avg       0.75      0.68      0.69       185
weighted avg       0.75      0.75      0.73       185

[[ 28  37]
 [  9 111]]
0.7513513513513513
GaussianNB()
              precision    recall  f1-score   support

           0       0.82      0.43      0.57        65
           1       0.75      0.95      0.84       120

    accuracy                           0.77       185
   macro avg       0.79      0.69      0.70       185
weighted avg       0.78      0.77      0.74       185

[[ 28  37]
 [  6 114]]
0.7675675675675676
SVC()
              precision    recall  f1-score   support

           0       0.96      0.42      0.58        65
           1       0.76      0.99      0.86       120

    accuracy                           0.79       185
  

**Evaluation metrics of a classification model,specifically a RandomForestClassifier() model, on a test dataset with 185 samples.**

**(1) precision:** Precision is the ratio of true positive predictions to all positive predictions made by the model. 
    In this case, precision for class 0 (the first row) is 0.86, which means that out of all the samples predicted as class 0       by the model, 86% were actually class 0
    Precision for class 1 (the second row) is 0.77, which means that out of all the samples predicted as class 1 by the model,     77% were actually class 1.    <br>

**(2) recall:** Recall is the ratio of true positive predictions to all actual positive samples in the dataset. 
    In this case, recall for class 0 is 0.46, which means that out of all the actual class 0 samples in the dataset, the model     correctly predicted 46%. Recall for class 1 is 0.96, which means that out of all the actual class 1 samples in the dataset,     the model correctly predicted 96%.    <br>

**(3) f1-score :** F1-score is a harmonic mean of precision and recall, and is a way to balance between precision and recall. 
    It is calculated as 2 x (precision x recall) / (precision + recall). 
    In this case, the f1-score for class 0 is 0.60, and the f1-score for class 1 is 0.85.   <br>

**(4) support :** The number of samples in each class in the test dataset.                  <br>

**(5) accuracy:** The overall accuracy of the model on the test dataset, which is the ratio of correct predictions to
      total number of predictions. In this case, the accuracy is 0.78.                      <br>

**(6) macro avg :** The macro average of precision, recall, and f1-score across both classes.  <br>

**(7) weighted avg :** The weighted average of precision, recall, and f1-score across both classes, taking into account the number of samples in each class.   <br>

**(8) [[30 35], [5 115]]:** The confusion matrix of the model on the test dataset, which shows the number of true positive, false positive, false negative, and true negative predictions for each class. In this case, the model correctly predicted 30 samples as class 0 and 115 samples as class 1, but incorrectly predicted 35 samples as class 0 and 5 samples as class 1. <br>

**(9) 0.7837837837837838 :** This is the accuracy of the model on the test dataset, which is the same as the accuracy mentioned earlier in the output.

# Conclusion: